src/CommitmentsV2/SFA.DAS.CommitmentsV2/Application/Commands/ApprenticeshipEmailAddressChangedByApprentice/ApprenticeshipEmailAddressChangedByApprenticeCommandHandler.cs

```diff
@@ -0,0 +1,71 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using MediatR;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Logging;
+using SFA.DAS.CommitmentsV2.Application.Commands.ApprenticeshipEmailAddressConfirmed;
+using SFA.DAS.CommitmentsV2.Data;
+using SFA.DAS.CommitmentsV2.Domain.Exceptions;
+using SFA.DAS.CommitmentsV2.Domain.Interfaces;
+using SFA.DAS.CommitmentsV2.Models.Api;
+using SFA.DAS.CommitmentsV2.Models.Api.Types;
+using SFA.DAS.CommitmentsV2.Shared.Extensions;
+using SFA.DAS.CommitmentsV2.Types;
+
+namespace SFA.DAS.CommitmentsV2.Application.Commands.ApprenticeshipEmailAddressChangedByApprentice
+{
+    public class ApprenticeshipEmailAddressChangedByApprenticeCommandHandler : AsyncRequestHandler<ApprenticeshipEmailAddressChangedByApprenticeCommand>
+    {
+        private readonly Lazy<ProviderCommitmentsDbContext> _db;
+        private readonly IApiClient _apimClient;
+        private readonly ILogger<ApprenticeshipEmailAddressChangedByApprenticeCommandHandler> _logger;
+
+        public ApprenticeshipEmailAddressChangedByApprenticeCommandHandler(Lazy<ProviderCommitmentsDbContext> db, IApiClient apimClient, ILogger<ApprenticeshipEmailAddressChangedByApprenticeCommandHandler> logger)
+        {
+            _db = db;
+            _apimClient = apimClient;
+            _logger = logger;
+        }
+
+        protected override async Task Handle(ApprenticeshipEmailAddressChangedByApprenticeCommand request, CancellationToken cancellationToken)
+        {
+            try
+            {
+                _logger.LogInformation("Getting Apprenticeship {0}", request.ApprenticeshipId);
+                var apprenticeshipTask = _db.Value.Apprenticeships.SingleAsync(a => a.Id == request.ApprenticeshipId, cancellationToken);
+
+                _logger.LogInformation("Getting Apprentice details for apprentice {0}", request.ApprenticeId);
+                var apprenticeTask = _apimClient.Get<ApprenticeResponse>(new GetApprenticeRequest(request.ApprenticeId));
+
+                await Task.WhenAll(apprenticeTask, apprenticeshipTask);
+
+                var apprenticeship = await apprenticeshipTask;
+                var apprentice = await apprenticeTask;
+
+                var status = apprenticeship.GetApprenticeshipStatus(DateTime.Now);
+
+                if (status != ApprenticeshipStatus.Stopped && status != ApprenticeshipStatus.Completed)
+                {
+                    if (apprenticeship.EmailAddressConfirmed == null)
+                    {
+                        throw new DomainException("Email",
```

Comment:
Is this an error condition, or a no-operation condition?  If it's a valid scenario it might be better to just log and ignore.

Reply:
I was thinking if the NSB messages are processed out of order, then we shouldn't update an email address which hasn't had the email address confirmed yet. This will fix itself when the EmailAddressConfirmedAddressEvent is processed.

If we ignore this the emails may get out of sync. Ie if we process these events in the wrong order